windows/src/engine/kmtip/tmgrsink.cpp

```diff
@@ -151,21 +151,51 @@ STDAPI CKMTipTextService::OnActivated(REFCLSID clsid, REFGUID guidProfile, BOOL
 {
   LogEnter();
 
-  guidActiveProfile = GUID_NULL;   // I4274
-
-  if(IsEqualGUID(clsid, c_clsidKMTipTextService))
-  {
-    SendDebugMessageFormat(L"CKMTipTextService::OnActivated(c_clsidKMTipTextService, <GUID>, %d)", fActivated);
-    if(fActivated) {
-      guidActiveProfile = guidProfile;
-      TIPNotifyActivate((GUID *)&guidProfile);
+  WCHAR bufClsid[40] = L"", bufProfile[40] = L"";
+  if (StringFromGUID2(clsid, bufClsid, 40) == 0) bufClsid[0] = 0;
+  if (StringFromGUID2(guidProfile, bufProfile, 40) == 0) bufProfile[0] = 0;
+  
+  if(IsEqualGUID(clsid, c_clsidKMTipTextService)) {
+    SendDebugMessageFormat(L"CKMTipTextService::OnActivated(c_clsidKMTipTextService, %s, %d)", bufProfile, fActivated);
+    TIPNotifyActivate(fActivated ? (GUID *)&guidProfile : NULL);
+  } else {
+    SendDebugMessageFormat(L"CKMTipTextService::OnActivated(%s, %s, %d)", bufClsid, bufProfile, fActivated);
+
+    if (fActivated) {
+      BOOL bIsDeactivating = TRUE;
+      ITfInputProcessorProfiles *pInputProcessorProfiles = NULL;
+
+      /* Look up the profile which is being activated, and only deactivate ourselves if it is a
+         Keyboard type TIP.
+      */
+      if (LogSUCCEEDED(CoCreateInstance(CLSID_TF_InputProcessorProfiles, NULL, CLSCTX_INPROC_SERVER,
+        IID_ITfInputProcessorProfiles, (void**)&pInputProcessorProfiles))) {
+        ITfInputProcessorProfileMgr *pInputProcessorProfileMgr = NULL;
+        if (LogSUCCEEDED(pInputProcessorProfiles->QueryInterface(&pInputProcessorProfileMgr))) {
+          IEnumTfInputProcessorProfiles *pEnumTfInputProcessorProfiles = NULL;
+          if (LogSUCCEEDED(pInputProcessorProfileMgr->EnumProfiles(0, &pEnumTfInputProcessorProfiles))) {
+            TF_INPUTPROCESSORPROFILE profile;
+            ULONG cFetch;
+            while (SUCCEEDED(pEnumTfInputProcessorProfiles->Next(1, &profile, &cFetch)) && cFetch == 1) {
+              if (guidProfile == profile.guidProfile) {
+                bIsDeactivating = profile.catid == GUID_TFCAT_TIP_KEYBOARD;
+                SendDebugMessageFormat(L"CKMTipTextService::OnActivated is it GUID_TFCAT_TIP_KEYBOARD? %d", bIsDeactivating);
+                break;
+              }
+            }
+            pEnumTfInputProcessorProfiles->Release();
+            pEnumTfInputProcessorProfiles = NULL;
+          }
+          pInputProcessorProfileMgr->Release();
+          pInputProcessorProfileMgr = NULL;
+        }
+        pInputProcessorProfiles->Release();
+        pInputProcessorProfiles = NULL;
+      }
+      if (bIsDeactivating) {
+        TIPNotifyActivate(NULL);
+      }
```

Comment:
Comparing to l. 160
```
TIPNotifyActivate(fActivated ? (GUID *)&guidProfile : NULL);
```

Is `TIPNotifyActivate(GUID *)&guidProfile)` not applicable to this block because it didn't match any profile, hence nothing to activate?

Reply:
If you look at l. 158:

```
if(IsEqualGUID(clsid, c_clsidKMTipTextService)) {
```

We only want to pass in the `guidProfile` that is being activated when it is a profile for the Keyman Text Service that is being activated. Any other keyboard-type profile (e.g. Chinese IME) needs Keyman to be disabled.

Reply:
I don't follow the big picture here. at l.160 if the clsid is our serivce then it needs to be activated or deactivted based on the `fActivated`.   Whereas by 196 we will not be activating and are only deactivating if the category tip of the text service is a keyboard OR if we can't access the profiles.  I say this because `bIsDeactivating  `is initialised to `TRUE`.
 I think the difference is calling TIPNotifyActive with NULL disables keyman. If we reach that if block I.196 we only want to disable because  either a) the porfile

Reply:
@rc-swag Yep, that's correct. Three(four) scenarios:

1. If another keyboard text service is activated, then we want to disable Keyman.
2. If Keyman text service is activated, then we want to enable Keyman.
3. If Keyman text service is deactivated, then we want to disable Keyman.
(4. if we are unable to enumerate profiles in scenario 1 above, then we want to disable Keyman to maintain existing behaviour. Error situation that's unlikely to occur.)

Ironically, I don't think `OnActivated` is ever called with `fActivated == FALSE` (for scenario 3). I've never seen it. But proper handling suggests we should do so.